vision/src/autogluon/vision/image_prediction/image_prediction.py

```diff
@@ -0,0 +1,319 @@
+        if self._classifier is None:
+            raise RuntimeError('Classifier is not initialized, try `fit` first.')
+        proba = self._classifier.predict(x)
+        if 'image' in proba.columns:
+            return proba.groupby(["image"]).agg(list)
+        return proba
+
+    def predict(self, x):
+        """Predict images as a whole, return labels(class category).
+
+        Parameters
+        ----------
+        x : str, pd.DataFrame or ndarray
+            The input, can be str(filepath), pd.DataFrame with 'image' column, or raw ndarray input.
+
+        Returns
+        -------
+
+        pd.DataFrame
+            The returned dataframe will contain labels. If more than one image in input,
+            the returned dataframe will contain `images` column, and all results are concatenated.
+        """
+        if self._classifier is None:
+            raise RuntimeError('Classifier is not initialized, try `fit` first.')
+        proba = self._classifier.predict(x)
+        if 'image' in proba.columns:
+            # multiple images
+            return proba.loc[proba.groupby(["image"])["score"].idxmax()].reset_index(drop=True)
+        else:
+            # single image
+            return proba.loc[[proba["score"].idxmax()]]
+
+    def predict_feature(self, x):
+        """Predict images visual feature representations, return the features as numpy (1xD) vector.
+
+        Parameters
+        ----------
+        x : str, pd.DataFrame or ndarray
+            The input, can be str(filepath), pd.DataFrame with 'image' column, or raw ndarray input.
+
+        Returns
+        -------
+
+        pd.DataFrame
+            The returned dataframe will contain image features. If more than one image in input,
+            the returned dataframe will contain `images` column, and all results are concatenated.
+        """
+        if self._classifier is None:
+            raise RuntimeError('Classifier is not initialized, try `fit` first.')
+        return self._classifier.predict_feature(x)
+
+    def evaluate(self, val_data):
+        """Evaluate model performance on validation data.
+
+        Parameters
+        ----------
+        val_data : pd.DataFrame or iterator
+            The validation data.
+        """
+        if self._classifier is None:
+            raise RuntimeError('Classifier not initialized, try `fit` first.')
+        return self._classifier.evaluate(val_data)
+
+    def fit_summary(self):
+        """Return summary of last `fit` process.
+
+        Returns
+        -------
+        dict
+            The summary of last `fit` process. Major keys are ('train_acc', 'val_acc', 'total_time',...)
+
+        """
+        return copy.copy(self._fit_summary)
+
+    def save(self, file_name):
+        """Dump predictor to disk.
+
+        Parameters
+        ----------
+        file_name : str
+            The file name of saved copy.
+
+        """
+        with open(file_name, 'wb') as fid:
+            pickle.dump(self, fid)
```

Comment:
Still confused how this doesn't cause issues. I thought mxnet models weren't pickle-able?

Reply:
mxnet models are generally not pickleable due to C pointers to ndarrays(especially cuda arrays, which are not even forkable), I use `__setstate__` and `__getstate__` to bypass the restriction by first use mxnet backend to dump the weights ,load it as binary, and finally remove the network object so the pickled object can be loaded back.